src/HydraAuction/Hydra/Monad.hs

```diff
@@ -0,0 +1,72 @@
+{-# OPTIONS_GHC -Wno-orphans #-}
+
+module HydraAuction.Hydra.Monad (
+  MonadHydra (..),
+  AwaitedHydraEvent (..),
+  waitForHydraEvent,
+  sendCommandAndWaitFor,
+) where
+
+-- Prelude imports
+import Prelude
+
+-- Haskell imports
+import Control.Monad (void)
+import GHC.Natural (Natural)
+import Unsafe.Coerce (unsafeCoerce)
+
+-- Cardano imports
+import Cardano.Api (AddressInEra (..))
+import Cardano.Api.UTxO qualified as UTxO
+
+-- Hydra imports
+import Hydra.Cardano.Api (Tx)
+import Hydra.Cardano.Api.Prelude (TxOut (..))
+
+-- HydraAuction imports
+import HydraAuction.Hydra.Interface (
+  HydraCommand (GetUTxO, NewTx),
+  HydraEvent (GetUTxOResponse, TxSeen),
+  HydraEventKind (GetUTxOResponseKind),
+ )
+import HydraAuctionUtils.Monads (MonadQueryUtxo (..), MonadSubmitTx (..), UtxoQuery (..))
+
+data AwaitedHydraEvent
+  = Any
+  | SpecificEvent HydraEvent
+  | SpecificKind HydraEventKind
+
+class Monad m => MonadHydra m where
+  sendCommand :: HydraCommand -> m ()
+  waitForHydraEvent' :: Natural -> AwaitedHydraEvent -> m HydraEvent
+
+defaultTimeout :: Natural
+defaultTimeout = 30
+
+waitForHydraEvent :: MonadHydra m => AwaitedHydraEvent -> m HydraEvent
+waitForHydraEvent = waitForHydraEvent' defaultTimeout
+
+sendCommandAndWaitFor :: MonadHydra m => AwaitedHydraEvent -> HydraCommand -> m HydraEvent
+sendCommandAndWaitFor awaitedSpec command = do
+  sendCommand command
+  waitForHydraEvent awaitedSpec
+
+instance (Monad m, MonadHydra m) => MonadSubmitTx m where
+  submitTx :: Tx -> m ()
+  submitTx = sendCommand . NewTx
+  awaitTx :: Tx -> m ()
+  awaitTx = void . waitForHydraEvent . SpecificEvent . TxSeen
+
+instance {-# OVERLAPPABLE #-} (Monad m, MonadHydra m) => MonadQueryUtxo m where
```

Comment:
I do not see which problems could OVERLAPPABLE introduce. Especially given that we do all instances in same package. And MLabs styleguide does not mention overlapping.  And we already use bunch of Orphan instances.

It builds now without OVELAPPABLE now though. I am now sure why, probably we do not import both instances in same file.

Reply:
There's a couple of reasons why this is bad: 
- behavior may change by omitting an instance declaration 
- especially in combination with orphan instances this is incredibly dangerous
- I think it's generally best practice to not reach for OVERLAPPING and OVERLAPPABLE without good reasons, I don't see a good reason here because we could just have a `default` impl or a `newtype` wrapper, which are very nice solutions 
- also, that it works at the moment is because you don't use the instance. overlapping and overlappable errors are discovered at call sites.

Reply:
> behavior may change by omitting an instance declaration

Please provide specific example how this will break our project.

> we could just have a default impl or a newtype wrapper, which are very nice solutions

Newtype is code cluttering hack, which I do not consider a good solution here. How this could be solved with default impls?

Reply:
Not sure how to deal with `newtype` for case of `* -> *` kinded types here. But you probably right that it is cleaner.

Please separate that to follow-up tech-debt issue.

Reply:
`newtype Wrap f a = Wrap (f a)`

Reply:
however, you can also do a `default` decl